lib/src/embedded/dispatcher.test.ts

```diff
@@ -0,0 +1,297 @@
+        expect(message.type).toEqual(
+          InboundMessage.MessageCase.CANONICALIZERESPONSE
+        );
+        done();
+      });
+
+      dispatcher.onCanonicalizeRequest(
+        () => new InboundMessage.CanonicalizeResponse()
+      );
+
+      const request = new OutboundMessage.CanonicalizeRequest();
+      request.setId(id);
+      outbound$.next({
+        payload: request,
+        type: OutboundMessage.MessageCase.CANONICALIZEREQUEST,
+      });
+    });
+
+    it('triggers the function call request callback', async done => {
+      const id = 1;
+
+      const dispatcher = new Dispatcher(outbound$, message => {
+        const expectedResponse = new InboundMessage.FunctionCallResponse();
+        expectedResponse.setId(id);
+        expect(message.payload).toEqual(expectedResponse);
+        expect(message.type).toEqual(
+          InboundMessage.MessageCase.FUNCTIONCALLRESPONSE
+        );
+        done();
+      });
+
+      dispatcher.onFunctionCallRequest(
+        () => new InboundMessage.FunctionCallResponse()
+      );
+
+      const request = new OutboundMessage.FunctionCallRequest();
+      request.setId(id);
+      outbound$.next({
+        payload: request,
+        type: OutboundMessage.MessageCase.FUNCTIONCALLREQUEST,
+      });
+    });
+  });
+
+  describe('multiple request listeners', () => {
+    it('supports multiple request callbacks', async done => {
+      const receivedRequests: number[] = [];
+
+      const dispatcher = new Dispatcher(outbound$, message => {
+        if (message.type === InboundMessage.MessageCase.FUNCTIONCALLRESPONSE) {
+          expect(receivedRequests).toEqual([1, 2]);
+          done();
+        }
+      });
+
+      dispatcher.onCanonicalizeRequest(request => {
+        receivedRequests.push(request.getId());
+        return new InboundMessage.CanonicalizeResponse();
+      });
+
+      dispatcher.onFunctionCallRequest(request => {
+        receivedRequests.push(request.getId());
+        return new InboundMessage.FunctionCallResponse();
+      });
+
+      const request = new OutboundMessage.CanonicalizeRequest();
+      request.setId(1);
+      outbound$.next({
+        payload: request,
+        type: OutboundMessage.MessageCase.CANONICALIZEREQUEST,
+      });
+
+      const request1 = new OutboundMessage.FunctionCallRequest();
+      request1.setId(2);
+      outbound$.next({
+        payload: request1,
+        type: OutboundMessage.MessageCase.FUNCTIONCALLREQUEST,
+      });
+    });
+  });
+
+  describe('protocol errors', () => {
+    beforeEach(() => {
+      dispatcher = new Dispatcher(outbound$, () => {});
+    });
+
+    it('throws if a request ID overlaps with that of an in-flight request', async done => {
+      dispatcher.error$.subscribe(
+        () => fail('expected error'),
+        error => {
+          expect(error.message).toBe(
+            'Request ID 0 is already in use by an in-flight request.'
+          );
+          done();
+        }
+      );
```

Comment:
Consider encapsulating this pattern in a helper function.

I wish you could just use the Dart test framework's stream testing APIs :disappointed:. On a cursory glance it doesn't look like RxJs has any equivalent, though.

Reply:
Yeah, it's a shame